README.md
# Aranea

Aranea is a gem to use in fault-tolerance testing. It simulates a remote dependency failing, so that you can see how your project reacts.

In Dungeons and Dragons, an Aranea is a giant spider that will ambush a group of adventurers using its illusion powers, then immobilize the strongest member, forcing the others to battle without a key ally.

Aranea currently only works with Faraday clients.

## Activating Aranea in a Deployed Rails or Rack Application

If Aranea has already been set up and configured to run in your testing environment, you can switch it on via a POST request to `/disable`. For example, in your shell, you could write

```sh
curl -d "" https://myapp-sandbox.example.com/disable?dependency=google
```

Your app should return the message "For the next 5 minutes, all requests to urls containing 'google' will 500". You can then see how your app would behave if google were down, without the bother of having to actually DOS google.

You must specify a dependency. If you want *all* external requests to fail, set `dependency=.`.  Any regular expression syntax will work, so if you want a subset of dependencies to fail, use a pipe-delimited list (`dependency=google|yahoo`).

You can override the duration of the test by specifying `minutes=` with a number from 1 to 60.

You can optionally specify the response body by setting `response=` and also the response header by setting `header=`. The endpoint to create a stubbed response uses query parameters, so your `response` and `header` fields must be URL encoded. For example, to return a response `{ "errors": "test_error" }`, you must send `response=%7B%20%22errors%22%3A%20%22test_error%22%20%7D`.

You can also override the simulated response code, if your app is meant to handle different failures differently. `?dependency=google&failure=404` will simulate a 404 (Not Found) response instead of a 500 (Internal Server Error). `?dependency=google&failure=timeout` will pretend the server never responded at all (although it will raise an error instantly; the illusion is not perfect).

### Sample calls and their effects

```
https://myapp-sandbox.example.com/disable?dependency=google
```
For the next 5 minutes, all requests to urls containing 'google' will 500.

---

```
https://myapp-sandbox.example.com/disable?dependency=.&failure=403
```
For the next 5 minutes, all external requests will 403.

---

```
https://myapp-sandbox.example.com/disable?dependency=sheep&failure=ssl_error
```
For the next 5 minutes, all external requests to urls containing `sheep` will raise a Faraday::SSLError.

---

```
https://myapp-sandbox.example.com/disable?dependency=google|yahoo&minutes=10
```
For the next 10 minutes, all requests to urls containing 'google' and/or 'yahoo' will 500.

---

```
https://myapp-sandbox.example.com/disable?dependency=google|yahoo&minutes=10&failure=timeout
```
For the next 10 minutes, all requests to urls containing 'google' and/or 'yahoo' will raise a Timeout error.

---

```
https://myapp-sandbox.example.com/disable?dependency=google|yahoo&minutes=10&failure=timeout&response=%7B%20%22errors%22%3A%20%22test_error%22%20%7D
```
Similar to the above example, except all external requests that match will now return `{ "errors": "test_error" }`.

---

```
https://myapp-sandbox.example.com/disable?dependency=google|yahoo&minutes=10&failure=timeout&response=%7B%20%22errors%22%3A%20%22test_error%22%20%7D&headers=%7B%20%22Content-Type%22%3A%20%22application%2Fjson%22%20%7D
```
Similar to the above example, except all external requests that match will now return `{ "errors": "test_error" }`, and will return the header `{ "Content-Type": "application/json" }`

## Activating Aranea Programmatically

From inside your application, you can run

```ruby
Aranea::Failure.create(
  pattern: /google/,
  minutes: 5,
  failure: 503,
  response_hash: { "errors": "test_error" },
  response_headers_hash: { "Content-Type": "application/json" }
)
```

All parameters are required in this form. They may alternatively be provided as strings ('google','5').

## Adding Aranea To Your Project

In your Gemfile:

```ruby
gem "aranea", "~> 0.1"
```

In your Faraday middleware stack (wherever and however you configure it):

```ruby
use Aranea::Faraday::FailureSimulator
```

Similarly, in your Rack middleware stack:

```ruby
if ENV['USE_ARANEA']
  use Aranea::Rack::FailureCreator
end
```

You'll want to ensure that Aranea (or at minimum the Rack middleware) is not active in production!

## Caveats and Limitations

Aranea needs to be included manually into every Faraday stack you have. If you miss one, or if you have a gem that makes http requests without using Faraday, your fault-tolerance test may falsely pass. Pull requests are welcome for compatibility with other http clients (aws-sdk is on the wishlist).

Failures are not necessarily shared between processes. Aranea will use `Rails.cache` if it exists and is available when it starts up, so if `Rails.cache` is shared between processes, a POST that hits one app worker will disable requests from all of them. But if you're not using Rails, your cache is in-memory, your cache is a file cache and your workers can be on different servers, or something else goes wrong, then some requests are likely to get through. In particular, jobs executed in the background may not hit the failure, even if they execute within the specified time frame.

## Owners

Aranea is (c) Medidata Solutions Worldwide and owned by whoever committed to it most recently. Original concept by Matthew Szenher.


CHANGELOG.md
# 0.2.0
- Allow Faraday 2.x.

# 0.1.0
- Require Ruby 2.5 and up.
- Require ActiveSupport 4.2 and up.
- Allow Faraday 1.x.
- Refresh code/specs and publish to rubygems.org.

# 0.0.3
- Added ability to specify the response body and response headers.

# 0.0.2
- Ability to raise Faraday::SSLError.

# 0.0.1
- Initial release.


spec/spec_helper.rb
# frozen_string_literal: true

require "active_support"
require "active_support/testing/time_helpers"
require "aranea"

RSpec.configure do |config|
  config.include ActiveSupport::Testing::TimeHelpers

  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

  # This option will default to `:apply_to_host_groups` in RSpec 4 (and will
  # have no way to turn it off -- the option exists only for backwards
  # compatibility in RSpec 3). It causes shared context metadata to be
  # inherited by the metadata hash of host groups and examples, rather than
  # triggering implicit auto-inclusion in groups with matching metadata.
  config.shared_context_metadata_behavior = :apply_to_host_groups

  # This allows you to limit a spec run to individual examples or groups
  # you care about by tagging them with `:focus` metadata. When nothing
  # is tagged with `:focus`, all examples get run. RSpec also provides
  # aliases for `it`, `describe`, and `context` that include `:focus`
  # metadata: `fit`, `fdescribe` and `fcontext`, respectively.
  config.filter_run_when_matching :focus

  # Allows RSpec to persist some state between runs in order to support
  # the `--only-failures` and `--next-failure` CLI options. We recommend
  # you configure your source control system to ignore this file.
  # config.example_status_persistence_file_path = "spec/examples.txt"

  # Limits the available syntax to the non-monkey patched syntax that is
  # recommended. For more details, see:
  #   - http://rspec.info/blog/2012/06/rspecs-new-expectation-syntax/
  #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://rspec.info/blog/2014/05/notable-changes-in-rspec-3/#zero-monkey-patching-mode
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = "doc"
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  # config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
end


spec/rack_middleware_spec.rb
# frozen_string_literal: true

RSpec.describe Aranea::Rack::FailureCreator do
  after do
    Aranea::Failure::Repository.clear
  end

  it "passes normal requests through transparently" do
    @app = double(:call)
    @env = { "REQUEST_METHOD" => "GET", "PATH_INFO" => "/widgets/1.json" }
    expect(@app).to receive(:call).with(@env.clone)

    described_class.new(@app).call(@env)
  end

  describe "handling POST requests to /disable" do
    before do
      allow_any_instance_of(described_class).to receive(:puts)

      @app = double(:call)
      @env = { "REQUEST_METHOD" => "POST", "PATH_INFO" => "/disable" }

      expect(@app).not_to receive(:call)
    end

    it "responds with a 201 (created) message" do
      @env["QUERY_STRING"] = "dependency=example"
      status, _headers, _response = described_class.new(@app).call(@env).flatten.to_a
      expect(status).to eq(201)
    end

    it "uses defaults when optional arguments are not provided" do
      @env["QUERY_STRING"] = "dependency=example"
      _status, _headers, response = described_class.new(@app).call(@env).flatten.to_a
      expect(response).to eq("For the next 5 minutes, all requests to urls containing 'example' will 500")
    end

    it "uses optional arguments when provided" do
      @env["QUERY_STRING"] = "dependency=example&minutes=10&failure=422&response=%7B%7D&headers=%7B%7D"
      _status, _headers, response = described_class.new(@app).call(@env).flatten.to_a
      expect(response).to eq("For the next 10 minutes, all requests to urls containing 'example' will 422")
    end

    it "accepts timeout as a failure" do
      @env["QUERY_STRING"] = "dependency=example&minutes=10&failure=timeout"
      _status, _headers, response = described_class.new(@app).call(@env).flatten.to_a
      expect(response).to eq("For the next 10 minutes, all requests to urls containing 'example' will timeout")
    end

    it "accepts ssl_error as a failure" do
      @env["QUERY_STRING"] = "dependency=example&minutes=10&failure=ssl_error"
      _status, _headers, response = described_class.new(@app).call(@env).flatten.to_a
      expect(response).to eq("For the next 10 minutes, all requests to urls containing 'example' will ssl_error")
    end

    it "is case insensitive when accepting string-based failures" do
      @env["QUERY_STRING"] = "dependency=example&minutes=10&failure=tImEoUt"
      _status, _headers, response = described_class.new(@app).call(@env).flatten.to_a
      expect(response).to eq("For the next 10 minutes, all requests to urls containing 'example' will timeout")
    end

    it "properly sets the expiry" do
      @env["QUERY_STRING"] = "dependency=example&minutes=10&failure=422"
      described_class.new(@app).call(@env)
      travel_to(Time.now + 6.minutes) do
        expect(Aranea::Failure::Repository.get).to be
      end
      travel_to(Time.now + 20.minutes) do
        expect(Aranea::Failure::Repository.get).not_to be
      end
    end

    it "responds with an error unless a dependency is provided" do
      @env["QUERY_STRING"] = ""
      _status, _headers, response = described_class.new(@app).call(@env).flatten.to_a
      expect(response).to eq("Please provide a dependency to simulate failing")
    end

    it "responds with an error on an unrecognized failure mode" do
      [200, 600, "llama", ""].each do |failure|
        @env["QUERY_STRING"] = "dependency=example&failure=#{failure}"
        _status, _headers, response = described_class.new(@app).call(@env).flatten.to_a
        expect(response).to eq("failure should be a 4xx or 5xx status code, timeout, or ssl_error; got #{failure}")
      end
    end

    it "responds with an error on a malformed minutes parameter" do
      [-1, 0.5, "llama", "", 61].each do |minutes|
        @env["QUERY_STRING"] = "dependency=example&minutes=#{minutes}"
        _status, _headers, response = described_class.new(@app).call(@env).flatten.to_a
        expect(response).to eq("minutes should be an integer from 1 to 60, got #{minutes}")
      end
    end

    it "responds with an error if a failure is already active" do
      @env["QUERY_STRING"] = "dependency=example&minutes=60"
      described_class.new(@app).call(@env).last
      status, _headers, response = described_class.new(@app).call(@env).flatten.to_a
      expect(status).to eq(422)
      expect(response).to match(
        %r{A failure is already in progress \(Failure on /example/ ending at approximately .+\).}
      )
    end

    it "catches and responds with any unexpected error" do
      allow_any_instance_of(described_class).to receive(:create_failure).and_raise("Out of cheese")
      @env["QUERY_STRING"] = "dependency=example"
      _status, _headers, response = described_class.new(@app).call(@env).flatten.to_a
      expect(response).to eq("Out of cheese")
    end
  end
end


spec/faraday_middleware_spec.rb
# frozen_string_literal: true

RSpec.describe Aranea::Faraday::FailureSimulator do
  before do
    @app = double(:call)
    @env = {}
  end

  after do
    Aranea::Failure::Repository.clear
  end

  context "no failure is active" do
    it "passes requests through transparently" do
      expect(@app).to receive(:call).with(@env.clone)
      described_class.new(@app).call(@env)
    end
  end

  context "a failure is active" do
    let(:pattern) { "yahoo|google" }
    let(:failure) { 415 }
    let(:minutes) { 100 }
    let(:response_hash) { { hello: "there" } }
    let(:response_headers_hash) { { "Content-type" => "application/json" } }

    before do
      Aranea::Failure.create(
        pattern: pattern,
        failure: failure,
        minutes: minutes,
        response_hash: response_hash,
        response_headers_hash: response_headers_hash
      )
    end

    context "the request does not match the specified pattern" do
      before do
        @env[:url] = "https://www.bing.com/search?q=adorable+kittens&go=&qs=n&form=QBLH&pq=adorable+kittens"
      end

      it "passes requests through transparently" do
        expect(@app).to receive(:call).with(@env.clone)
        described_class.new(@app).call(@env)
      end
    end

    context "the request matches the specified pattern" do
      before do
        allow_any_instance_of(described_class).to receive(:puts)

        @env[:url] = "https://www.google.com/search?q=adorable+kittens"
        expect(@app).not_to receive(:call)
      end

      it "blocks the request from completing" do
        described_class.new(@app).call(@env)
      end

      it "returns the specified response code, headers, and body" do
        response = described_class.new(@app).call(@env)
        expect(response.status).to eq(415)
        expect(response.body).to eq(response_hash.to_json)
        expect(response.headers).to include(response_headers_hash)
      end
    end
  end

  context "a timeout is simulated" do
    before do
      Aranea::Failure.create(pattern: "yahoo|google", failure: "timeout", minutes: 100)
      @env[:url] = "https://www.google.com/search?q=adorable+kittens"
      allow_any_instance_of(described_class).to receive(:puts)
    end

    it "throws a Faraday::TimeoutError" do
      expect { described_class.new(@app).call(@env) }.to raise_error(Faraday::TimeoutError)
    end
  end

  context "when an SSL error is simulated" do
    before do
      Aranea::Failure.create(pattern: "yahoo|google", failure: "ssl_error", minutes: 100)
      @env[:url] = "https://www.google.com/search?q=adorable+puppies"
      allow_any_instance_of(described_class).to receive(:puts)
    end

    it "throws a Faraday::SSLError" do
      expect { described_class.new(@app).call(@env) }.to raise_error(Faraday::SSLError)
    end
  end
end


lib/aranea/failure_repository.rb
# frozen_string_literal: true

module Aranea
  # TODO: Look into moving whitelisting of consumer hostnames to here and allowing it to be configurable
  # via the consuming application

  class Failure
    class << self
      def current
        retrieve_failure || NullFailure.new
      end

      def create(params)
        failure = retrieve_failure
        raise FailureFailure, "A failure is already in progress (#{failure})." if failure

        failure = new(params)
        Repository.store(failure, params[:minutes].minutes)
        failure
      end

      def retrieve_failure
        Repository.get
      end
    end

    attr_accessor :expiration_date

    def initialize(params)
      @pattern = Regexp.new(params[:pattern])
      @response = params[:failure]
      @response_hash = params[:response_hash] || {}
      @response_headers_hash = params[:response_headers_hash] || {}
    end

    def should_fail?(request_env, _app)
      @pattern.match(request_env[:url].to_s)
    end

    def respond!
      case @response
      when "timeout"
        raise ::Faraday::TimeoutError, "Fake failure from Aranea"
      when "ssl_error"
        raise ::Faraday::SSLError, "Fake failure from Aranea"
      else
        ::Faraday::Response.new(
          status: @response.to_i,
          body: @response_hash.to_json,
          response_headers: @response_headers_hash
        )
      end
    end

    def to_s
      "Failure on #{@pattern.inspect} ending at approximately #{@expiration_date}"
    end

    # TODO: Actually implement Repository pattern, dependency injection and all.
    # As is we only support sharing between multiple instances if Rails.cache exists and does
    class Repository
      KEY = "aranea_current_failure"

      class << self
        def store(failure, lifespan)
          failure.expiration_date = Time.now + lifespan
          cache.write(KEY, failure, expires_in: lifespan)
        end

        def get
          cache.read(KEY)
        end

        def clear
          cache.delete(KEY)
        end

        def cache
          @cache ||= rails_cache || memory_store
        end

        def rails_cache
          defined?(Rails.cache) && Rails.cache
        end

        def memory_store
          require "active_support/cache"
          ActiveSupport::Cache::MemoryStore.new
        end
      end
    end
  end

  class NullFailure
    def should_fail?(_request_env, _app)
      false
    end
  end
end


lib/aranea/rack/aranea_middleware.rb
# frozen_string_literal: true

module Aranea
  FailureFailure = Class.new(RuntimeError)

  ALLOWED_MINUTES = (1..60).freeze

  module Rack
    class FailureCreator
      def initialize(app, config = {})
        @app = app
        @config = config
      end

      def call(env)
        if failure_creation_request?(env)
          response = begin
            [create_failure(::Rack::Utils.parse_query(env["QUERY_STRING"])), 201, {}]
          rescue FailureFailure => e
            [e.message, 422, {}]
          rescue => e
            [e.message, 500, {}]
          end
          ::Rack::Response.new(*response).finish
        else
          @app.call(env)
        end
      end

      protected

      def failure_creation_request?(env)
        env["REQUEST_METHOD"] == "POST" && env["PATH_INFO"] == "/disable"
      end

      def create_failure(options)
        dependency = options.fetch("dependency") do
          raise FailureFailure, "Please provide a dependency to simulate failing"
        end
        minutes    = options.fetch("minutes", 5)
        failure    = options.fetch("failure", "500")
        response   = options.fetch("response", "{}") # Expect URI encoded
        headers    = options.fetch("headers", "{}")

        # These will emit a parsing error if the response/headers URI encoded string is malformed.
        response_hash = JSON.parse(CGI.unescape(response)).to_hash
        response_headers_hash = JSON.parse(CGI.unescape(headers)).to_hash

        unless /\A((4|5)\d\d|timeout|ssl_error)\Z/i.match?(failure)
          raise FailureFailure, "failure should be a 4xx or 5xx status code, timeout, or ssl_error; got #{failure}"
        end

        unless ALLOWED_MINUTES.cover?(minutes.to_i)
          raise FailureFailure,
            "minutes should be an integer from #{ALLOWED_MINUTES.begin} to #{ALLOWED_MINUTES.end}, got #{minutes}"
        end

        Failure.create(
          pattern: dependency,
          minutes: minutes.to_i,
          failure: failure,
          response_hash: response_hash,
          response_headers_hash: response_headers_hash
        )

        result = "For the next #{minutes} minutes, all requests to urls containing '#{dependency}' " \
                 "will #{failure.downcase}"

        # TODO: injectable logger
        puts "Aranea: #{result}"

        result
      end
    end
  end
end
